In [4]:
!pip install numpy
!pip install pandas


[notice] A new release of pip is available: 24.1 -> 24.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 24.1 -> 24.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import numpy as np
import pandas as pd

In [6]:
data = pd.read_csv('laptop_price.csv', encoding='latin-1')

In [7]:
data.head()

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   
3          4   Apple  MacBook Pro  Ultrabook    15.4   
4          5   Apple  MacBook Pro  Ultrabook    13.3   

                     ScreenResolution                         Cpu   Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz   8GB   
1                            1440x900        Intel Core i5 1.8GHz   8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz   8GB   
3  IPS Panel Retina Display 2880x1800        Intel Core i7 2.7GHz  16GB   
4  IPS Panel Retina Display 2560x1600        Intel Core i5 3.1GHz   8GB   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94  
2         Intel HD Graphics 620  No OS  1.86kg       575.00  
3            AMD Radeon Pro 455  macOS  1.83kg      2537.45  
4  Intel Iris Plus Graphics 650  macOS  1.37kg      1803.60

Data Analaysis

In [8]:
data.shape

(1303, 12)

In [9]:
data.isnull().sum()

laptop_ID           0
Company             0
Product             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Gpu                 0
OpSys               0
Weight              0
Price_euros         0
dtype: int64

//No any missing values in any column

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         1303 non-null   int64  
 1   Company           1303 non-null   object 
 2   Product           1303 non-null   object 
 3   TypeName          1303 non-null   object 
 4   Inches            1303 non-null   float64
 5   ScreenResolution  1303 non-null   object 
 6   Cpu               1303 non-null   object 
 7   Ram               1303 non-null   object 
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight            1303 non-null   object 
 11  Price_euros       1303 non-null   float64
dtypes: float64(2), int64(1), object(9)
memory usage: 122.3+ KB


note - maching learning model can read only integer, float type data types.

data.head(2)

// then remove GB and kg in Ram and Weight columns to make them int and float

In [11]:
data['Ram'] = data['Ram'].str.replace('GB','').astype('int32')
data['Weight'] = data['Weight'].str.replace('kg','').astype('float32')

In [12]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69  
1        Intel HD Graphics 6000  macOS    1.34       898.94

Company column

In [13]:
data['Company'].value_counts()

Company
Dell         297
Lenovo       297
HP           274
Asus         158
Acer         103
MSI           54
Toshiba       48
Apple         21
Samsung        9
Mediacom       7
Razer          7
Microsoft      6
Vero           4
Xiaomi         4
Chuwi          3
Fujitsu        3
Google         3
LG             3
Huawei         2
Name: count, dtype: int64

In [14]:
len(data['Company'].value_counts())

19

In [15]:
def add_company(input):
    if input== 'Samsung' or input== 'Mediacom' or input== 'Razer' or input== 'Microsoft' or input== 'Vero' or input== 'Xiaomi' or input== 'Chuwi' or input== 'Fujitsu' or input== 'Google' or input== 'LG' or input== 'Huawei':
        return 'Other'
    else:
        return input

In [16]:
data['Company'] = data['Company'].apply(add_company)

In [17]:
data['Company'].value_counts()

Company
Dell       297
Lenovo     297
HP         274
Asus       158
Acer       103
MSI         54
Other       51
Toshiba     48
Apple       21
Name: count, dtype: int64

Product Column

In [18]:
data['Product'].value_counts()

Product
XPS 13                                 30
Inspiron 3567                          29
250 G6                                 21
Vostro 3568                            19
Legion Y520-15IKBN                     19
                                       ..
L403NA-GA013TS (N3350/4GB/32GB/W10)     1
IdeaPad 500-15ISK                       1
GP62M 7RDX                              1
V110-15ISK (i3-6006U/4GB/1TB/No         1
15-BA015wm (E2-7110/4GB/500GB/W10)      1
Name: count, Length: 618, dtype: int64

In [19]:
len(data['Product'].value_counts())

618

// Need to remove Product column cuz it has 618 raws

TypeName Column

In [20]:
data['TypeName'].value_counts()

TypeName
Notebook              727
Gaming                205
Ultrabook             196
2 in 1 Convertible    121
Workstation            29
Netbook                25
Name: count, dtype: int64

In [21]:
len(data['TypeName'].value_counts())

6

ScreenResolution Column

In [22]:
data['ScreenResolution'].value_counts()

ScreenResolution
Full HD 1920x1080                                507
1366x768                                         281
IPS Panel Full HD 1920x1080                      230
IPS Panel Full HD / Touchscreen 1920x1080         53
Full HD / Touchscreen 1920x1080                   47
1600x900                                          23
Touchscreen 1366x768                              16
Quad HD+ / Touchscreen 3200x1800                  15
IPS Panel 4K Ultra HD 3840x2160                   12
IPS Panel 4K Ultra HD / Touchscreen 3840x2160     11
4K Ultra HD / Touchscreen 3840x2160               10
IPS Panel 1366x768                                 7
Touchscreen 2560x1440                              7
4K Ultra HD 3840x2160                              7
IPS Panel Retina Display 2304x1440                 6
IPS Panel Retina Display 2560x1600                 6
Touchscreen 2256x1504                              6
IPS Panel Quad HD+ / Touchscreen 3200x1800         6
IPS Panel Touchscreen 2560x14

In [23]:
data['Touchscreen'] = data['ScreenResolution'].apply(lambda x:1 if 'Touchscreen' in x else 0)
data['Ips'] = data['ScreenResolution'].apply(lambda x:1 if 'IPS' in x else 0)

In [24]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69            0    1  
1        Intel HD Graphics 6000  macOS    1.34       898.94            0    0

CPU column

In [25]:
data['Cpu'].value_counts()

Cpu
Intel Core i5 7200U 2.5GHz       190
Intel Core i7 7700HQ 2.8GHz      146
Intel Core i7 7500U 2.7GHz       134
Intel Core i7 8550U 1.8GHz        73
Intel Core i5 8250U 1.6GHz        72
                                ... 
Intel Core i5 7200U 2.70GHz        1
Intel Core M M7-6Y75 1.2GHz        1
Intel Core M 6Y54 1.1GHz           1
AMD E-Series 9000 2.2GHz           1
Samsung Cortex A72&A53 2.0GHz      1
Name: count, Length: 118, dtype: int64

In [26]:
len(data['Cpu'].value_counts())

118

In [27]:
data['cpu_name'] = data['Cpu'].apply(lambda x:" ".join(x.split()[0:3]))

In [28]:
data['cpu_name'].value_counts()

cpu_name
Intel Core i7               527
Intel Core i5               423
Intel Core i3               136
Intel Celeron Dual           80
Intel Pentium Quad           27
Intel Core M                 19
AMD A9-Series 9420           12
AMD A6-Series 9220            8
Intel Celeron Quad            8
AMD A12-Series 9720P          7
Intel Atom x5-Z8350           5
AMD A8-Series 7410            4
Intel Atom x5-Z8550           4
AMD A9-Series 9410            3
Intel Pentium Dual            3
AMD Ryzen 1700                3
AMD A9-Series A9-9420         2
AMD E-Series E2-9000e         2
AMD A10-Series A10-9620P      2
AMD A6-Series A6-9220         2
AMD E-Series 7110             2
AMD A10-Series 9620P          2
AMD A10-Series 9600P          2
Intel Xeon E3-1505M           2
Intel Xeon E3-1535M           2
Intel Atom X5-Z8350           2
Intel Atom x5-Z8300           1
AMD E-Series 6110             1
AMD E-Series 9000e            1
AMD E-Series E2-6110          1
AMD FX 9830P                  1

In [29]:
def new_processor(name):
     if name == 'Intel Core i7' or name == 'Intel Core i5' or name == 'Intel Core i3':
         return name
     else:
         if name.split()[0] == 'AMD':
             return 'AMD'
         else:
             return 'Other'

In [30]:
data['cpu_name'] = data['cpu_name'].apply(new_processor)

In [31]:
data['cpu_name'].value_counts()

cpu_name
Intel Core i7    527
Intel Core i5    423
Other            155
Intel Core i3    136
AMD               62
Name: count, dtype: int64

GPU column

In [32]:
data['Gpu'].value_counts()

Gpu
Intel HD Graphics 620      281
Intel HD Graphics 520      185
Intel UHD Graphics 620      68
Nvidia GeForce GTX 1050     66
Nvidia GeForce GTX 1060     48
                          ... 
Nvidia Quadro M500M          1
AMD Radeon R7 M360           1
Nvidia Quadro M3000M         1
Nvidia GeForce 960M          1
ARM Mali T860 MP4            1
Name: count, Length: 110, dtype: int64

In [33]:
len(data['Gpu'].value_counts())

110

In [34]:
data['gpu_name'] = data['Gpu'].apply(lambda x:" ".join(x.split()[0:1])) 

In [35]:

data['gpu_name'].value_counts()

gpu_name
Intel     722
Nvidia    400
AMD       180
ARM         1
Name: count, dtype: int64

In [36]:
data.shape

(1303, 16)

// Deleting the ARM row

In [37]:
data = data[data['gpu_name'] != 'ARM']

In [38]:
data.shape

(1302, 16)

In [39]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69            0    1   
1        Intel HD Graphics 6000  macOS    1.34       898.94            0    0   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel

OpSys Column

In [40]:
data['OpSys'].value_counts()

OpSys
Windows 10      1072
No OS             66
Linux             62
Windows 7         45
Chrome OS         26
macOS             13
Mac OS X           8
Windows 10 S       8
Android            2
Name: count, dtype: int64

In [41]:
def new_os(name):
    if name == 'Windows 10' or name == 'Windows 7' or name == 'Windows 10 S':
        return 'Windows'
    elif name == 'macOS' or name == 'Mac OS X':
        return 'Mac'
    elif name == 'Linux':
        return name
    else:
        return 'Other'

In [42]:
data['OpSys'] = data['OpSys'].apply(new_os)

In [43]:
data['OpSys'].value_counts()

OpSys
Windows    1125
Other        94
Linux        62
Mac          21
Name: count, dtype: int64

In [44]:
data.head(2)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   

                     ScreenResolution                   Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600  Intel Core i5 2.3GHz    8   
1                            1440x900  Intel Core i5 1.8GHz    8   

                            Gpu OpSys  Weight  Price_euros  Touchscreen  Ips  \
0  Intel Iris Plus Graphics 640   Mac    1.37      1339.69            0    1   
1        Intel HD Graphics 6000   Mac    1.34       898.94            0    0   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel

// Drop unnecessary columns

In [45]:
data = data.drop(columns = ['laptop_ID', 'Inches', 'Product', 'ScreenResolution', 'Cpu', 'Gpu'])

In [46]:
data.head()

Company   TypeName  Ram  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0   Apple  Ultrabook    8    Mac    1.37      1339.69            0    1   
1   Apple  Ultrabook    8    Mac    1.34       898.94            0    0   
2      HP   Notebook    8  Other    1.86       575.00            0    0   
3   Apple  Ultrabook   16    Mac    1.83      2537.45            0    1   
4   Apple  Ultrabook    8    Mac    1.37      1803.60            0    1   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel  
2  Intel Core i5    Intel  
3  Intel Core i7      AMD  
4  Intel Core i5    Intel

// Then do One Hot Encoding

In [47]:
data = pd.get_dummies(data)

In [48]:
data.head()

Ram  Weight  Price_euros  Touchscreen  Ips  Company_Acer  Company_Apple  \
0    8    1.37      1339.69            0    1         False           True   
1    8    1.34       898.94            0    0         False           True   
2    8    1.86       575.00            0    0         False          False   
3   16    1.83      2537.45            0    1         False           True   
4    8    1.37      1803.60            0    1         False           True   

   Company_Asus  Company_Dell  Company_HP  ...  OpSys_Other  OpSys_Windows  \
0         False         False       False  ...        False          False   
1         False         False       False  ...        False          False   
2         False         False        True  ...         True          False   
3         False         False       False  ...        False          False   
4         False         False       False  ...        False          False   

   cpu_name_AMD  cpu_name_Intel Core i3  cpu_name_Intel Core i5  \
0         False                   False                    True   
1         False                   False                    True   
2         False                   False                    True   
3         False                   False                   False   
4         False                   False                    True   

   cpu_name_Intel Core i7  cpu_name_Other  gpu_name_AMD  gpu_name_Intel  \
0                   False           False         False            True   
1                   False           False         False            True   
2                   False           False         False            True   
3                    True           False          True           False   
4                   False           False         False            True   

   gpu_name_Nvidia  
0            False  
1            False  
2            False  
3            False  
4            False  

[5 rows x 32 columns]

In [49]:
data.shape

(1302, 32)

// Label column having price values. not a categorical variable. continuous values are there. So this is a regression problem.

Model Building and Selection

In [50]:
X = data.drop('Price_euros', axis=1)
y = data['Price_euros']

// in x, axis = 1 naming is to specify x as a column. x is having all the data except Price_euros.

In [54]:
!pip install scikit-learn


[notice] A new release of pip is available: 24.1 -> 24.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Divide the data set for training and testing

In [56]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25)

In [57]:
X_train.shape, X_test.shape

((976, 31), (326, 31))

Model Training

In [58]:
def model_accuracy(model):
    model.fit(X_train, y_train)
    accuracy = model.score(X_test, y_test)
    print(str(model)+ ' ---> ' +str(accuracy))

In [59]:
from sklearn.linear_model import LinearRegression
linearReg = LinearRegression()
model_accuracy(linearReg)

LinearRegression() ---> 0.6966613125864745


In [61]:
from sklearn.linear_model import Lasso
lasso = Lasso()
model_accuracy(lasso)

Lasso() ---> 0.6957956348910967


In [62]:
from sklearn.tree import DecisionTreeRegressor
decisionTree = DecisionTreeRegressor()
model_accuracy(decisionTree)

DecisionTreeRegressor() ---> 0.7150012417495304


In [63]:
from sklearn.ensemble import RandomForestRegressor
randomForest = RandomForestRegressor()
model_accuracy(randomForest)

RandomForestRegressor() ---> 0.7887785435289869


// RandomForestRegressor has the highest accuracy

Hyperparameter Tunning

In [65]:
from sklearn.model_selection import GridSearchCV

parameters = {'n_estimators':[10, 50, 100],
              'criterion':['squared_error','absolute_error','poisson']}

grid_obj = GridSearchCV(estimator=randomForest, param_grid=parameters)

grid_fit = grid_obj.fit(X_train, y_train)

best_model = grid_fit.best_estimator_

best_model

RandomForestRegressor(criterion='absolute_error')

In [67]:
best_model.score(X_test, y_test)

0.7942131518876137

In [68]:
X_test.columns

Index(['Ram', 'Weight', 'Touchscreen', 'Ips', 'Company_Acer', 'Company_Apple',
       'Company_Asus', 'Company_Dell', 'Company_HP', 'Company_Lenovo',
       'Company_MSI', 'Company_Other', 'Company_Toshiba',
       'TypeName_2 in 1 Convertible', 'TypeName_Gaming', 'TypeName_Netbook',
       'TypeName_Notebook', 'TypeName_Ultrabook', 'TypeName_Workstation',
       'OpSys_Linux', 'OpSys_Mac', 'OpSys_Other', 'OpSys_Windows',
       'cpu_name_AMD', 'cpu_name_Intel Core i3', 'cpu_name_Intel Core i5',
       'cpu_name_Intel Core i7', 'cpu_name_Other', 'gpu_name_AMD',
       'gpu_name_Intel', 'gpu_name_Nvidia'],
      dtype='object')

Model Saving

In [69]:
import pickle
with open('predictor.pickle', 'wb') as file:
    pickle.dump(best_model, file)

In [71]:
predicted_value = best_model.predict([[8, 1.3, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0]])
predicted_value

C:\Users\DELL\Desktop\Laptop Price Predictor\model\env\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


array([1256.7772])